File: ios/Classes/handlers/Deeplink/ApphudDeeplinkBridge.swift
//
//  ApphudDeeplinkBridge.swift
//  apphud
//
//  Bridges the native deep link attribution handler to Flutter over the
//  `apphud/deeplink` method channel.
//

import Foundation
import ApphudSDK

public class ApphudDeeplinkBridge: NSObject, FlutterPlugin {

    /// Keeps the last bridge that enabled a Flutter deeplink handler so we can
    /// re-apply it after `Apphud.start` / `startManually`, which reset the
    /// native handler to `nil` when no `deeplinkHandler` argument is passed.
    private static weak var activeBridge: ApphudDeeplinkBridge?
    private static var handlerEnabled = false

    private var channel: FlutterMethodChannel

    internal init(channel: FlutterMethodChannel) {
        self.channel = channel
    }

    public static func register(with registrar: FlutterPluginRegistrar) {}

    /// Re-installs the Flutter deeplink handler if Dart previously enabled it.
    /// Call after `Apphud.start` / `startManually` (those APIs clear the handler).
    @MainActor
    static func reapplyHandlerIfNeeded() {
        guard handlerEnabled, let bridge = activeBridge else { return }
        bridge.installHandler()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "setDeeplinkHandler":
            let enabled = (args?["enabled"] as? Bool) ?? false
            setDeeplinkHandler(enabled: enabled)
            result(nil)
        case "requestDeferredDeeplinkAttribution":
            requestDeferredDeeplinkAttribution()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func setDeeplinkHandler(enabled: Bool) {
        Self.handlerEnabled = enabled
        if enabled {
            Self.activeBridge = self
            Task { @MainActor in
                self.installHandler()
            }
        } else {
            if Self.activeBridge === self {
                Self.activeBridge = nil
            }
            Task { @MainActor in
                Apphud.setDeeplinkHandler(nil)
            }
        }
    }

    @MainActor
    private func installHandler() {
        Apphud.setDeeplinkHandler { [weak self] attribution, kind, url in
            self?.notify(attribution: attribution, kind: kind, url: url)
        }
    }

    private func requestDeferredDeeplinkAttribution() {
        Task { @MainActor in
            Apphud.requestDeferredDeeplinkAttribution()
        }
    }

    private func notify(attribution: [String: Any],
                        kind: ApphudDeeplinkAttributionKind,
                        url: URL?) {
        let kindString: String
        switch kind {
        case .deferred:
            kindString = "deferred"
        default:
            kindString = "direct"
        }

        let arguments: [String: Any?] = [
            "attribution": attribution,
            "kind": kindString,
            "url": url?.absoluteString,
        ]

        // Apphud may invoke the deep link handler from a background thread
        // (especially for deferred attribution). Flutter platform channels must
        // be called from the main thread.
        if Thread.isMainThread {
            channel.invokeMethod("onDeeplinkAttribution", arguments: arguments)
        } else {
            DispatchQueue.main.async { [channel] in
                channel.invokeMethod("onDeeplinkAttribution", arguments: arguments)
            }
        }
    }
}

File: android/src/main/kotlin/com/apphud/fluttersdk/handlers/ApphudDeeplinkBridge.kt
package com.apphud.fluttersdk.handlers

import android.app.Activity
import com.apphud.sdk.Apphud
import com.apphud.sdk.ApphudDeeplinkAttributionKind
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Bridges the native deep link attribution handler to Flutter over the
 * `apphud/deeplink` method channel.
 *
 * Direct deep links are fed to the native SDK by the host app's Activity (via
 * `Apphud.handleIntent(intent)`); this bridge only forwards the resulting
 * attribution back to Flutter and triggers deferred attribution.
 */
class ApphudDeeplinkBridge(handleOnMainThreadP: HandleOnMainThread) :
    MethodChannel.MethodCallHandler {

    private var handleOnMainThread = handleOnMainThreadP
    private var channel: MethodChannel? = null
    var activity: Activity? = null

    fun setMethodCallHandler(channel: MethodChannel?) {
        this.channel?.setMethodCallHandler(null)
        if (channel == null) {
            this.channel = null
        } else {
            this.channel = channel
            this.channel!!.setMethodCallHandler(this)
        }
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setDeeplinkHandler" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                setDeeplinkHandler(enabled)
                result.success(null)
            }

            "requestDeferredDeeplinkAttribution" -> {
                requestDeferredDeeplinkAttribution(result)
            }

            else -> result.notImplemented()
        }
    }

    private fun setDeeplinkHandler(enabled: Boolean) {
        if (enabled) {
            Apphud.setDeeplinkHandler { attribution, kind, uri ->
                val kindString = when (kind) {
                    ApphudDeeplinkAttributionKind.DEFERRED -> "deferred"
                    else -> "direct"
                }
                val arguments = hashMapOf<String, Any?>(
                    "attribution" to attribution,
                    "kind" to kindString,
                    "url" to uri?.toString(),
                )
                handleOnMainThread {
                    channel?.invokeMethod("onDeeplinkAttribution", arguments)
                }
            }
        } else {
            Apphud.setDeeplinkHandler(null)
        }
    }

    private fun requestDeferredDeeplinkAttribution(result: MethodChannel.Result) {
        val activity = this.activity
        if (activity == null) {
            handleOnMainThread {
                result.error(
                    "no_activity",
                    "Deferred deep link attribution requires an attached Activity. " +
                        "Call requestDeferredDeeplinkAttribution() while the app is in the foreground.",
                    null,
                )
            }
            return
        }
        Apphud.requestDeferredDeeplinkAttribution(activity)
        handleOnMainThread { result.success(null) }
    }
}
